### ISSUE TITLE
Slow highlighting and type inference.

### ISSUE THREAD

**Issue by davenza** [intellij-rust/intellij-rust#2486]
Hi,

I've been using IntelliJ-Rust for months but the highlighting and type inference works really slow in the last version for me.

This is an screenshot:

![intellij](https://user-images.githubusercontent.com/25250234/39098670-aaaccc60-466e-11e8-90e4-4eb45fa1bdbf.png)

As you can see, the highlighting is not complete. For example, the #[macro_use] at the start or the product() function. Furthermore, the partial highlighting was very slow. It lasted minutes to highlight those 3 functions. While I am writing this issue (5-10 minutes), the #[macro_use] got highlighted. There is no type inference as well. I thought about trying the previous version of the plugin (to be sure that is the plugin's fault), but I don't know how to install a previous version. I also tried with the nightly version of the plugin without better luck.

CLion 2018.1.1
IntelliJ-Rust 0.2.0.2093-181

I have reproduced this issue in other projects. In fact, with larger files, the highlighting seems even slower. I don't know what additional information to provide or how to reproduce the issue.

**Comment by Undin** [comment]
Looks like reason of slow highlighting is macro expansion that was introduced in the latest release
We know that macro expansion can significantly slow some IDE features and we are working on its improvement.
Try to switch off this feature (`Languages & Frameworks > Rust > Expand macros`) and check again.

**Comment by Undin** [comment]
cc @vlad20012

**Comment by davenza** [comment]
@Undin Disabling macro expansion does not solve the problem. I think it does the partial highlighting a bit faster (yet slower than in previous versions) in some files. However, in some files  it just does not finish to process them. Testing with the file in the screenshot (58 lines), gave me the same results that enabling macro expansion. My impression is that the plugin suddenly freezes, and stops highlighting. I discovered that clicking or putting the cursor in a function name, sometimes (not always) makes the plugin "wake up" and highlights the selected function.


I've just tested that I have the same issue in another pc (i7 6700k. I think it is not a bad CPU ). Same configuration of CLion and plugin.

**Comment by Undin** [comment]
@davenza Hm, it's weird
Could you provide your project (if it's open source, of course)?

To install the previous version of the plugin you should download it (for example, from plugin repository: https://plugins.jetbrains.com/plugin/8182-rust) and install via `Install plugin from disk...` action

**Comment by davenza** [comment]
@Undin Of course. You can find the project at https://github.com/davenza/Python_Rust_benchmark.  
The file in the screenshot is located at benches/bench_main. Of course, the code compiles gracefully with cargo bench.

But I had problems with other projects such as https://github.com/bluss/petgraph. For example, in src/graph_impl/mod.rs


I just (re)-checked that the previous version (0.2.0.2092-181) does **NOT** have this issue.

**Comment by vasily-kirichenko** [comment]
It seems to affect syntax highlighting only, because current symbol highlighting work OK:

![image](https://user-images.githubusercontent.com/873919/39142834-7d4941c4-4734-11e8-8bef-c2c2020460b9.png)

**Comment by davenza** [comment]
> What's interesting is that there is no significant CPU load while the code is not highlighted.

@vasily-kirichenko Yes, there is no CPU load. That is the reason why I  had the impression that the plugin "freezes". 

> It seems to affect syntax highlighting only, because current symbol highlighting work OK:

I don't know if that happens to you, but sometimes (not always) while doing symbol highlighting in a method that does not have correct syntax highlighting, the syntax highlighting re-runs correctly.

And the type inference does not work until all the highlighting is correct.

**Comment by kimond** [comment]
I'm experiencing the same issue having slow highlights and "freezes".

IntelliJ IDEA 2018.1.5 (Ultimate Edition)
Build #IU-181.5281.24, built on June 12, 2018
JRE: 1.8.0_152-release-1248-b6 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.17.2-1-ARCH

Rust plugin version: 0.2.0.2096-181

![selection_466](https://user-images.githubusercontent.com/1423906/41514041-ba71ef70-7271-11e8-97f1-2505c3f111b6.png)

**Comment by davenza** [comment]
With the last version of CLion and Rust plugin (0.2.0.2097-181), it just don't work anymore:

CLion 2018.1.5
Build #CL-181.5281.33, built on June 14, 2018
JRE: 1.8.0_152-release-1136-b39 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.13.0-45-generic

The plugin remains frozen from the start for me, so I think is more than a performance issue (or it is even slower every day).

![frozen](https://user-images.githubusercontent.com/25250234/41707315-19b31602-753e-11e8-8f53-d1ecfd3fd5d0.png)

**Comment by Undin** [comment]
@davenza We've just released a new plugin version with some performance improvements. Could you please check it and give us feedback?
It's not expected that it significantly reduces time of initial file highlighting but we expect better performance and lower cpu consuming while typing

**Comment by davenza** [comment]
@Undin For me is working pretty well. It completes the syntax highlighting and type inference for any file in less than a second even with the expand macro option enabled.
